server/data_tiledb/tiledb_engine.py

```diff
@@ -0,0 +1,288 @@
+            'sm.num_reader_threads': 32,
+        })
+
+    def update(self, location=None, config=None):
+        self.config = config
+        if location:
+            self.url = location
+            if self.url[-1] != '/':
+                self.url += '/'
+
+            self._validate_and_initialize()
+
+    @staticmethod
+    def pre_checks(location):
+        if not TileDbEngine.isvalid(location):
+            raise RuntimeError(f"tiledb matrix is not valid: {location}")
+
+    @staticmethod
+    def file_size(location):
+        return 0
+
+    @staticmethod
+    def open(location, args):
+        return TileDbEngine(location, args)
+
+    def get_location(self):
+        return self.url
+
+    def get_name(self):
+        return "cellxgene TileDb engine version"
+
+    def get_library_versions(self):
+        return dict(tiledb=tiledb.__version__)
+
+    def get_path(self, *urls):
+        return path_join(self.url, *urls)
+
+    def lsuri(self, uri):
+        """
+        given a URI, do a tiledb.ls but normalizing for all path weirdness:
+        * S3 URIs require trailing slash.  file: doesn't care.
+        * results on S3 *have* a trailing slash, Posix does not.
+
+        returns list of (absolute paths, type) *without* trailing slash
+        in the path.
+        """
+        def _cleanpath(p):
+            if p[-1] == '/':
+                return p[:-1]
+            else:
+                return p
+
+        if uri[-1] != '/':
+            uri += '/'
+
+        result = []
+        tiledb.ls(uri,
+                  lambda path, type: result.append((_cleanpath(path), type)),
+                  ctx=self.tiledb_ctx)
+        return result
+
+    @staticmethod
+    def isvalid(url):
+        """
+        Return True if this looks like a valid CXG, False if not.  Just a quick/cheap
+        test, not to be fully trusted.
+        """
+        if not tiledb.object_type(url) == "group":
+            return False
+        if not tiledb.object_type(path_join(url, "obs")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "var")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "X")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "emb")) == "group":
+            return False
+        return True
+
+    def _validate_and_initialize(self):
+        if not self.isvalid(self.url):
+            raise RuntimeError(f"invalid tiledb dataset {self.url}")
+
+    def get_array(self, name, items=None):
+        p = self.get_path(name)
+        if items is None:
+            items = slice(None)
+        with tiledb.DenseArray(p, mode="r", ctx=self.tiledb_ctx) as A:
+            data = A[items]
+        return data
+
+    def get_embedding_array(self, ename, items=None):
+        return self.get_array(f"emb/{ename}", items)
+
+    def get_X_array(self, obs_items=None, var_items=None):
```

Comment:
remember that tiledb expects multi_index to use inclusive indexing for index ranges/slices, and that it is faster if you collapse consecutive integer indices into slices (or tuples).

Reply:
Done